File: Backend/src/main/java/com/Trabajo_Final_Beltran/service/DescuentoService.java

package com.Trabajo_Final_Beltran.service;

import com.Trabajo_Final_Beltran.dto.request.CreateDescuentoRequest;
import com.Trabajo_Final_Beltran.dto.request.UpdateDescuentoRequest;
import com.Trabajo_Final_Beltran.dto.response.DescuentoResponse;
import com.Trabajo_Final_Beltran.dto.response.MensajeResponse;

import java.util.List;

public interface DescuentoService {

    DescuentoResponse crearDescuento(CreateDescuentoRequest request);

    DescuentoResponse editarDescuento(Long id, UpdateDescuentoRequest request);

    List<DescuentoResponse> listarDescuentos();

    DescuentoResponse obtenerPorId(Long id);

    MensajeResponse activar(Long id);

    MensajeResponse desactivar(Long id);
}

File: frontEnd/src/app/core/services/descuento.service.ts
import { HttpClient } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { DescuentoResponse } from '../models/descuento-response';
import { CreateDescuentoRequest } from '../models/create-descuento-request';
import { UpdateDescuentoRequest } from '../models/update-descuento-request';

export interface MensajeResponse {
  mensaje: string;
}

@Injectable({
  providedIn: 'root'
})
export class DescuentoService {
  private readonly http = inject(HttpClient);
  private readonly apiUrl = `${environment.baseUrl}/descuentos`;

  listarDescuentos(): Observable<DescuentoResponse[]> {
    return this.http.get<DescuentoResponse[]>(this.apiUrl);
  }

  obtenerDescuentoPorId(id: number): Observable<DescuentoResponse> {
    return this.http.get<DescuentoResponse>(`${this.apiUrl}/${id}`);
  }

  crearDescuento(request: CreateDescuentoRequest): Observable<DescuentoResponse> {
    return this.http.post<DescuentoResponse>(this.apiUrl, request);
  }

  editarDescuento(id: number, request: UpdateDescuentoRequest): Observable<DescuentoResponse> {
    return this.http.put<DescuentoResponse>(`${this.apiUrl}/${id}`, request);
  }

  activarDescuento(id: number): Observable<MensajeResponse> {
    return this.http.patch<MensajeResponse>(`${this.apiUrl}/${id}/activar`, null);
  }

  desactivarDescuento(id: number): Observable<MensajeResponse> {
    return this.http.patch<MensajeResponse>(`${this.apiUrl}/${id}/desactivar`, null);
  }
}
